src/org/rascalmpl/library/lang/paths/Windows.rsc

```diff
@@ -0,0 +1,113 @@
+    = current: "."
+    | parent : ".."
+    | name   : PathChar+ \ ".." \ "."
+    ;
+
+lexical Drive = [A-Za-z];
+
+lexical Slashes = Slash+ !>> [\\/];
+
+lexical Slash = [\\/];
+
+lexical WindowsFilePath = {PathSegment Slashes}* segments Slashes? [\ .] !<< (); // only the last segment must not end in spaces.
+
+import ParseTree;
+
+@synopsis{Convert a windows path literal to a source location URI}
+@description{
+1. parses the path using the grammar for ((WindowsPath))
+2. takes the _literal_ name components using string interpolation `"<segment>"`. This means no decoding/encoding happens at all while extracting
+hostname, share name and path segment names. Also all superfluous path separators are skipped.
+3. uses `loc + str` path concatenation with its builtin character encoding to construct the URI. Also
+the right path separators are introduced. 
+}
+loc parseWindowsPath(str input, loc src=|unknown:///|) = mapPathToLoc(parse(#WindowsPath, input, src));
+
+@synopsis{UNC}
+loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><PathChar* hostName><Slashes _><PathChar* shareName><Slashes _><WindowsFilePath path>`)
+    = (|unc://<hostName>/| + "<shareName>" | it + "<segment>" | segment <- path.segments );
+
+@synopsis{Absolute: given the drive and relative to its root.}
+loc mapPathToLoc((WindowsPath) `<Drive drive>:<Slashes _><WindowsFilePath path>`) 
+    = (|file:///<drive>:/| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Drive relative: relative to the current working directory on the given drive.}
+loc mapPathToLoc((WindowsPath) `<Drive drive>:<WindowsFilePath path>`) 
+    = (|file:///<drive>:.| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Directory relative: relative to the root of the current drive.}
+loc mapPathToLoc((WindowsPath) `<Slash _><WindowsFilePath path>`) 
+    = (|cwdrive:///| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Relative to the current working directory on the current drive.}
+loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) 
+    = (|cwd:///| | it + "<segment>" | segment <- path.segments);
+
+private bool IS_WINDOWS = /win/i := getSystemProperty("os.name");
+
+test bool uncSharePath()
+    = parseWindowsPath("\\\\Server2\\Share\\Test\\Foo.txt")
+    == |unc://Server2/Share/Test/Foo.txt|;
+
+test bool uncDrivePath()
+    = parseWindowsPath("\\\\system07\\C$\\")
+    == |unc://system07/C$|;
+
+test bool simpleDrivePathC()
+    = parseWindowsPath("C:\\Program Files\\Rascal")
+    == |file:///C:/Program%20Files/Rascal|;
+
+test bool mixedSlashesDrivePathC()
+    = parseWindowsPath("C:\\Program Files/Rascal")
+    == |file:///C:/Program%20Files/Rascal|;
+
+test bool trailingSlashesDrivePathC()
+    = parseWindowsPath("C:\\Program Files\\Rascal\\\\")
+    == |file:///C:/Program%20Files/Rascal|;
+
+test bool simpleDrivePathD()
+    = parseWindowsPath("D:\\Program Files\\Rascal")
+    == |file:///D:/Program%20Files/Rascal|;
+
+test bool uncNetworkShareOk() {
```

Comment:
if the tests are in this file, are they even tested by our regular JUnitRunners?